# SingleCellRNAseqTools
A collection of scripts for processing single-cell RNA-seq read data, including protocols that use single-cell library prep kits to sequence barcoded pools of thousands of cells.
##GetGeneCountsFromThreePrimeWellOverloadMIT.py
For some experiments, there is an interest in understanding expression variation among cell populations assigned to various conditions, but the amount of RNA obtained from those populations is too small to serve as input to standard bulk RNA-seq library preparation protocols. In such cases, and where there is less interest in cell-to-cell variation, one can use single-cell library preparation methods to build libraries from pools of cells. MIT's BioMicroCenter currently uses the protocol of Soumillion et al. (http://biorxiv.org/content/early/2014/03/05/003236) to generate such heaviliy 3'-end biased single-end RNA-seq libraries. Fastq files generated by this protocol produce reads with headers that contain (a) an index (identifying a set of plate rows) and a well id within those rows that together correspond to a specific sequenced pool of cells, and (b) a UMI. Under the assumption that only a UMI should only map to the same start location in a genome once, GetGeneCountsFromThreePrimeWellOverloadMIT.py takes a bwa-generated bam file and a bed file of gene bondaries for the reference genome, and extracts mRNA counts per gene.
